### ISSUE TITLE
make genius more robust to multiple queries

### ISSUE THREAD

**Issue by tsoukinator** [JosiahParry/genius#52]
I have noticed when I am querying using possible_album, my returned results often consist of songs without lyrics at all.
This appears to affect random depending on the time I run the query, and I am unable to determine why.

`sixteen_stone <- genius::possible_album(artist = "Bush", album = "Sixteen Stone", info = "all")`

When comparing two files run with the same query, it was actually a different selection of songs which had their lyrics downloaded, both times.

Edit: To provide clarity - the problem is songs which should have lyrics, instead return a single empty row.

**Comment by JosiahParry** [comment]
Using your example, all tracks return song lyrics using both `genius_album()` and `possible_album()`. Can you provide a reproducible example?

**Comment by tsoukinator** [comment]
I currently have 400 csv files containing lyrics of artists, where some songs are retrieved as a single row, with no information.
I used the above example. Is there some form of rate limiting if we don't use our own API Key?

See the bottom rows in this dataset - this csv contains my results of the above query - without any further manipulation post retrieving the data via the API wrapper.
[](https://drive.google.com/file/d/1y21BZ37aoo-e243ZttI8QY7MUWiIGHnv/view?usp=sharing)

**Comment by JosiahParry** [comment]
@tsoukinator, `genius` does not use the Genius API as the API cannot provide the lyrics. 

Since the code provided works on my machine with the CRAN released version of genius I cannot help solve this issue. 
If you can provide a reproducible example I'd love to help.

Below is my reprex in which the tracks you have associated NAs have the correct  lyrics. 

``` r

ga <- genius::genius_album(artist = "Bush", album = "Sixteen Stone", info = "all")
#> Joining, by = c("album_name", "track_n", "track_url", "track_title")
dplyr::count(ga, track_title)
#> # A tibble: 13 x 2
#>    track_title             n
#>    <chr>               <int>
#>  1 Alien                  48
#>  2 Body                   36
#>  3 Bomb                   36
#>  4 Comedown               90
#>  5 Comedown (Acoustic)    38
#>  6 Everything Zen         45
#>  7 Glycerine              38
#>  8 Little Things          56
#>  9 Machinehead            43
#> 10 Monkey                 43
#> 11 Swim                   42
#> 12 Testosterone           69
#> 13 X-Girlfriend            1
```

<sup>Created on 2020-07-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>


Perhaps you were rate limited by the browser—though I've never had this happen for me. 
I would recommend building rate limiting into your fetching of lyrics. 

something along the lines of 

```
f <- function(...) {
    Sys.sleep(round(runif(1, 0, 10)))
    genius_album(...)
}
```

**Comment by tsoukinator** [comment]
Thanks for the suggestion - I've actually been using your package since
early last year, where I was able to fetch 1400 artist lyric discographies
without hassle.
It's puzzling to know that the query is working on your side but not mine -
if that's the case, I'll have to continue troubleshooting with rate
limiting and different machines/networks to see if I can find anything else
on my side which might be related.

Thanks for making this package available to the world. Hopefully I can get
it functional again soon, as I've got big plans for it.

**Comment by tsoukinator** [comment]
The one weird thing that I have noticed - the songs which fail to retrieve
lyrics are not the same every time. Which is very puzzling.

**Comment by JosiahParry** [comment]
If the songs are failing randomly, I suspect your browser is being rate limited. Try adding in a random sleep to it. 

I would love to learn more about what you've got going on! 

Again, if you can provide any additional context as to when this happens and with what tracks that would be great.

If you're on a windows machine the encoding might be causing problems for you. This has been the case in the past. 

One thing you can do is use `gen_album_url()` or `gen_song_url()` to check if the urls that are being creates are the correct ones for you. 

Alternatively you can use `genius_tracklist()` to check all of the urls that are provided when running `genius_album()` (these are more trustworthy than doing it song by song). 

Furthermore, if you're getting the lyrics for many tracks or albums, I'd recommend you use `add_genius()` instead as the `possible_*()` functions are meant mainly for internal and developmental use. 

Perhaps this short case study would be useful https://mirr.netlify.app/ .

**Comment by JosiahParry** [comment]
Things I can do to make this better:

- reset cookies with each song
- randomize the user-agent for each song
- add optional sleeps.

**Comment by tsoukinator** [comment]
Thanks for all the information Josiah - I'll have to start my
troubleshooting over the weekend.

I'm aiming big right now - my main goal is to allow for vocab/linguistic
analyses on artists based on song lyrics (to pair with the kind of metrics
Spotify normally provides).
The end result, would ideally be a dashboarding tool which allows a user to
take a deeper look into many aspects of an artist - whether that be song
metrics, lyrics, sales/estimated earnings - although for now, I'm focussing
primarily on strengthening my database on the lyric side of things.

It's small steps at a time - and as much of my ideas rely upon the lyrics,
if there's a gap here, everything else I build will suffer in quality.
Will experiment with retries/sleep operations - I also need them for the
ngramr package I'm also using.

**Comment by tsoukinator** [comment]
An update on my troubleshooting - I'm trying using just the standard
genius_lyrics operator. In this situation, I cannot retrieve the lyrics for
Glycerine, but can retrieve the lyrics for Bomb.
song2 <- genius::genius_lyrics(artist = "Bush", song = "glycerine", info =
"simple")
song2

song <- genius::genius_lyrics(artist = "Bush", song = "Bomb", info =
"simple")
song

---
I then tried the following functions together, I am able to retrieve the
lyrics.

glycerine <- genius::gen_song_url(artist = "Bush", song = "Glycerine")
genius::genius_url(url=glycerine, info = "simple")

---
However, I went back to this function which previously worked, to retrieve
the track Bomb, and this data frame came back with no rows.
song <- genius::genius_lyrics(artist = "Bush", song = "Bomb", info =
"simple")

Very puzzled at what is going on.

**Comment by tsoukinator** [comment]
Another piece of progression - I also tried using the package geniusr .....
I'm encountering the same kind of issues.
geniusr::get_lyrics_search(artist_name = "Bush", song_title = "Bomb")

Would suggest there's something with the API itself - potentially genius
themselves have changed something.

**Comment by tsoukinator** [comment]
Last update, threw the get_lyrics function into a loop to test whether data
would be returned or not (using the same song as the search point).

Unsure why this is happening for me and you haven't experienced it, but the
results here seem to repeat what I saw in my previous tests - gaps in data
being retrieved.

artist_name <- "Bush"
song_name <- "Glycerine"

for (i in 1:50) {
lyrics <- try(genius::genius_lyrics(artist = artist_name, song = song_name,
info = "simple"));

if(nrow(lyrics) == 0){
    print(paste("Iteration :",i,". No Data",sep=""))
} else {
    print(paste("Iteration :",i,". Data Found",sep=""))
}
}
---------------------------------------
[1] "Iteration :1. No Data"
[1] "Iteration :2. No Data"
[1] "Iteration :3. No Data"
[1] "Iteration :4. No Data"
[1] "Iteration :5. No Data"
[1] "Iteration :6. Data Found"
[1] "Iteration :7. No Data"
[1] "Iteration :8. Data Found"
[1] "Iteration :9. No Data"
[1] "Iteration :10. Data Found"
[1] "Iteration :11. No Data"
[1] "Iteration :12. Data Found"
[1] "Iteration :13. No Data"
[1] "Iteration :14. No Data"
[1] "Iteration :15. Data Found"
[1] "Iteration :16. No Data"
[1] "Iteration :17. Data Found"
[1] "Iteration :18. No Data"
[1] "Iteration :19. No Data"
[1] "Iteration :20. Data Found"
[1] "Iteration :21. Data Found"
[1] "Iteration :22. No Data"
[1] "Iteration :23. No Data"
[1] "Iteration :24. No Data"
[1] "Iteration :25. No Data"
[1] "Iteration :26. Data Found"
[1] "Iteration :27. Data Found"
[1] "Iteration :28. Data Found"
[1] "Iteration :29. No Data"

**Comment by lm-create** [comment]
Hi guys,

I'm super new to R but I seem to be having a similar issue. I am successfully getting lyrics of some songs on an album, but half or more are just appearing as a one line entry with no lyrics. This happens whether I use genius_album or possible_album.

I tried this code for the song I'm working with:
geniusr::get_lyrics_search(artist_name = "INXS", song_title = "New Sensation")

and it works. However, the lyrics for this song don't appear in the results when I run this:
lyricsKick2 <- possible_album(artist = "INXS", album = "Kick")

When I run this:
lyricsKick <- possible_album(artist = "INXS", album = "Kick", info="simple")
I have 12 songs with missing lyrics

lyricsKick2 <- possible_album(artist = "INXS", album = "Kick")
I have 6 songs with missing lyrics
I ran it a second time and had 10 songs with missing lyrics

Strangely, one of the 6 missing songs in the second example does appear with lyrics in the first example. I also ran the exact same Bush and Glycerine code by Anthony and got a different set of results for "No Data" v "Date Found"

I'm guessing this all points to rate limiting as the issue? I'm on a Mac.

I don't suppose you can point me in the right direction with adding something to address the rate limiting? Josiah - you mentioned this, but I wasn't sure what to put in the brackets (as I said, I'm super new to this!)?
f <- function(...) {
    Sys.sleep(round(runif(1, 0, 10)))
    genius_album(...)
}

Thanks so much!

**Comment by tsoukinator** [comment]
Hello there new person - thanks for your interest - it's sparked me into trying to resolve this one for good.

I ended up writing a larger than expected workaround to solve for this problem, I've pasted it below - feel free to substitute your values in at **artist_name** and **album_name** to use for your own queries!

```
# Package and Parameter Setup
###############################################
library(genius) # Josiahparry's genius R API Wrapper package
library(dplyr) # Filter and data transformation
library(data.table) # Rbind function

# User Parameters
###############################################
artist_name <- "Bush" # Type in your artist name here
album_name <- "Sixteen Stone" # Type in your album name here

sleep_time <- 5 # Time to sleep in seconds between iterations retrieving NA lyrics

# Additional configuration pre-run
###############################################
iterations_to_fix <- 0 # Counter to keep track of iterations requires to fix data

# Create dataframe to hold missing songs - designed with the info = all argument in mind.
missing_song_df <- data.frame(album_name = character(), track_n = integer(), artist = character(), track_title = character(), line = integer(), lyric = character(), element = character(), element_artist = character())

# Start lyric repair function
###############################################

# Try initial lyric search
lyrics <- try(genius::genius_album(artist = artist_name, album = album_name, info = "all"));

# Find track names with 0 lyrics
tracks_na <- unique(filter(lyrics, is.na(artist))$track_title)
# Find track numbers with 0 lyrics (for appending to the missing song df later)
track_n_na <- unique(filter(lyrics, is.na(artist))$track_n)

if(length(tracks_na) > 0){
  
  print(paste(length(tracks_na)," songs failed to retrive lyrics: ",sep=""))
  print("Starting query loop.")

while (length(tracks_na) > 0 ) {
  # Count iterations to fix
  iterations_to_fix <- iterations_to_fix + 1
  
  print(paste("Iteration: ",iterations_to_fix,". Sleeping for: ",sleep_time," seconds before retry.",sep=""))
  Sys.sleep(sleep_time)
  
  # Remove NAs from original dataset
  lyrics <- filter(lyrics, !is.na(artist))
  
  for (t in 1:length(tracks_na)) {
    
    print(paste("Iteration ",iterations_to_fix,". Song ",t,"/",length(tracks_na),". Searching again for song name: ",tracks_na[t],", artist: ",artist_name,".",sep=""))
    
    missing_song_lyric <- genius::genius_lyrics(artist = artist_name, song = tracks_na[t], info = "all")
    
    # If the query returns NA again, we need to handle this as an error and try again next iteration
    if (nrow(missing_song_lyric) == 0) {
        print(paste("Song: ",tracks_na[t],", - Requery failed. Will attempt on next iteration.",sep=""))
      
        # Create empty df row to re-process through this function later
        missing_song_lyric <- data.frame(album_name = album_name, track_n = track_n_na[t], artist = artist_name, track_title = tracks_na[t], line = 0, lyric = "NA", element = "NA", element_artist = "")
        
        missing_song_df <- rbind(missing_song_df,missing_song_lyric)
      
    } else {
          # Add album name to df - (only comes with the album query, not lyric query)
          print(missing_song_lyric$track_title[1])
          missing_song_lyric$album_name <- album_name
          # Add track number to df (doesn't come with the lyric track query) - (only comes with the album query, not lyric query)
          missing_song_lyric$track_n <- track_n_na[t]
          # Combine missing lyric dfs together
          missing_song_df <- rbind(missing_song_df,missing_song_lyric)
    }

  }
    # Combine total re-queried lyrics from run
    lyrics <- rbind(lyrics,missing_song_df)
  
    print(paste(length(tracks_na)," tracks were re-queried.",sep=""))
    tracks_na <- unique(filter(lyrics, is.na(artist))$track_title)
    track_n_na <- unique(filter(lyrics, is.na(artist))$track_n)
    print(paste(length(tracks_na)," tracks still require lyrics.",sep=""))
  # Test to see if any NAs remain
}

}

print(paste("Songs attached to original df. It took ",iterations_to_fix," iteration(s) to resolve.",sep=""))

# Re-order by track number
lyrics <- lyrics[order(lyrics$track_n),]

# Verification: Show track numbers (should now be complete sequential list)
unique(lyrics$track_n)

# Display completed lyrics
lyrics

```

**Comment by lm-create** [comment]
@tsoukinator wow, that's super impressive, thank you!

It definitely yielded the most lyrics yet, but there were 3 songs where the requery failed and they are still entered as one of those 1 row for the entire song entries.

Lauren

**Comment by tsoukinator** [comment]
Hello there new person - thanks for your interest - it's sparked me into trying to resolve this one for good.

I ended up writing a larger than expected workaround to solve for this problem, I've pasted it below - feel free to substitute your values in at artist_name and album_name to use for your own queries!

```
# Package and Parameter Setup
###############################################
library(genius) # Josiahparry's genius R API Wrapper package
library(dplyr) # Filter and data transformation
library(data.table) # Rbind function

# User Parameters
###############################################
artist_name <- "Bush" # Type in your artist name here
album_name <- "Sixteen Stone" # Type in your album name here

sleep_time <- 30 # Time to sleep in seconds between iterations retrieving NA lyrics

# Additional configuration pre-run
###############################################
iterations_to_fix <- 0 # Counter to keep track of iterations requires to fix data

# Start lyric repair function
###############################################

# Try initial lyric search
lyrics <- try(genius::genius_album(artist = artist_name, album = album_name, info = "all"));

# Find track names with 0 lyrics
tracks_na <- unique(filter(lyrics, is.na(artist))$track_title)

if(length(tracks_na) > 0){
  
  print(paste(length(tracks_na)," songs failed to retrive lyrics: ",sep=""))
  print("Starting query loop.")

while (length(tracks_na) > 0 ) {
  # Count iterations to fix
  iterations_to_fix <- iterations_to_fix + 1
  
  # Create dataframe to hold missing songs - designed with the info = all argument in mind.
  missing_song_df <- data.frame(album_name = character(), track_n = integer(), artist = character(), track_title = character(), line = integer(), lyric = character(), element = character(), element_artist = character())

  print(paste("Iteration: ",iterations_to_fix,". Sleeping for: ",sleep_time," seconds before retry.",sep=""))
  Sys.sleep(sleep_time)
  
  # Find track names with 0 lyrics
  tracks_na <- unique(filter(lyrics, is.na(artist))$track_title)
  # Find track numbers with 0 lyrics (for appending to the missing song df later)
  track_n_na <- unique(filter(lyrics, is.na(artist))$track_n)
    
  # Remove NAs from original dataset
  lyrics <- filter(lyrics, !is.na(artist))
  
  for (t in 1:length(tracks_na)) {
    
    print(paste("Iteration ",iterations_to_fix,". Song ",t,"/",length(tracks_na),". Searching again for song name: ",tracks_na[t],", artist: ",artist_name,".",sep=""))
    
    missing_song_lyric <- genius::genius_lyrics(artist = artist_name, song = tracks_na[t], info = "all")
    
    # If the query returns NA again, we need to handle this as an error and try again next iteration
    if (nrow(missing_song_lyric) == 0) {
        print(paste("Song: ",tracks_na[t],", - Requery failed. Will attempt on next iteration.",sep=""))
      
        # Create empty df row to re-process through this function later
        missing_song_lyric <- data.frame(album_name = album_name, track_n = track_n_na[t], artist = NA, track_title = tracks_na[t], line = 0, lyric = "NA", element = "NA", element_artist = "")
        
        missing_song_df <- rbind(missing_song_df,missing_song_lyric)
      
    } else {
          # Add album name to df - (only comes with the album query, not lyric query)
          print(paste("Retrieved song: ",tracks_na[t]," successfully.",sep=""))
          missing_song_lyric$album_name <- album_name
          # Add track number to df (doesn't come with the lyric track query) - (only comes with the album query, not lyric query)
          missing_song_lyric$track_n <- track_n_na[t]
          # Combine missing lyric dfs together
          missing_song_df <- rbind(missing_song_df,missing_song_lyric)
    }

  }
    # Combine total re-queried lyrics from run
    lyrics <- rbind(lyrics,missing_song_df)
    
    # Test to see if any NAs remain
    print(paste(length(tracks_na)," tracks were re-queried.",sep=""))
    tracks_na <- unique(filter(lyrics, is.na(artist))$track_title)
    track_n_na <- unique(filter(lyrics, is.na(artist))$track_n)
    print(paste(length(tracks_na)," tracks still require lyrics.",sep=""))
    
}

}

print(paste("Songs attached to original df. It took ",iterations_to_fix," iteration(s) to resolve.",sep=""))

# Re-order by track number
lyrics <- lyrics[order(lyrics$track_n),]

# Verification: Show track numbers (should now be complete sequential list)
unique(lyrics$track_n)

# Display completed lyrics
lyrics

```

**Comment by tsoukinator** [comment]
> @tsoukinator wow, that's super impressive, thank you!
> 
> It definitely yielded the most lyrics yet, but there were 3 songs where the requery failed and they are still entered as one of those 1 row for the entire song entries.
> 
> Lauren

Hey Lauren - I had a couple of errors in my last post - please take the code I most recently posted above!
Thanks :)

**Comment by JosiahParry** [comment]
Have either of you used add_genius() as suggested? It is unclear what the
issue is here. My guess is that you’re sending too many requests too
quickly and your IP is being blocked. Please try implementing a Sys.sleep()
call between requests.

Please try the method illustrated in mirr.netlify.com as well.

**Comment by tsoukinator** [comment]
Thanks for the suggestion Josiah - I have now also implemented a parameter in my previous post, to allow for a sleep time between individual track requests, as well as album retries. In this scenario, the album sleep time is likely no longer needed.

Add_genius may have saved me some time had I looked at it sooner. Will have to keep that in mind for future builds. For now, my above code seems to do the trick.

**Comment by lm-create** [comment]
@tsoukinator amazing! That worked this time and I could see it kept retrying each track until it was successful. Thank you so much.

Thank you @JosiahParry ... add_genius is a bit outside my ability right now (I've only been using R for a couple of weeks) but I'll keep persevering as I have a lot of lyrics I'll be needing! I'm so grateful for the genius package - it's a lifesaver for me.

**Comment by tsoukinator** [comment]
@lm-create - Thanks, I'm glad you found it useful. It's honestly been a thorn in my side over the last few weeks since I got back into doing my data hobby projects.

From @JosiahParry 's responses, it seems in the long run, building a solution to run purely off genius_lyrics to get one track at a time (with sleep periods in between) may be better - although that does require prior knowledge of the song names you require.

~ If you're like me and you also discover the **spotifyr** package, you can more easily get an artist's discography via the **get_artist_audio_features()** function. This will give you all the song/album names you could be after - and actually will also give you a whole bunch of extra metrics on the songs via Spotify which can be really interesting!

**Comment by lm-create** [comment]
@tsoukinator always nice to resolve those thorns! :)

Yes I love spotifyr!

**Comment by JosiahParry** [comment]
@lm-create I'd suggest taking a look at `add_genius()` it's a whole lot easier than writing your own for loop! :) 
As noted in the README, I've built a tutorial. You can access it via the tutorial pane in RStudio (version 1.3) or by running `learnr::run_tutorial("genius_tutorial", "genius")` in the console.

@tsoukinator If you're using {spotifyr} already, I'd suggested using the function `get_discography()` i built it for that purpose :P 

Again I'd recommend giving `mirr.netlfiy.app` a perusal. For example there is a handy function definition in there for getting track level audio features without needing the track ID. (https://mirr.netlify.app/audio-features-spotify.html).

**Comment by lm-create** [comment]
@JosiahParry brilliant, thank you so much - those are really helpful suggestions and I'll definitely get stuck into them!

**Comment by JosiahParry** [comment]
Action Items from this issue:

- build in option sleep timer for rate limiting
- delete cookies for each new session
- set user-agents

**Comment by tsoukinator** [comment]
> @lm-create I'd suggest taking a look at `add_genius()` it's a whole lot easier than writing your own for loop! :)
> As noted in the README, I've built a tutorial. You can access it via the tutorial pane in RStudio (version 1.3) or by running `learnr::run_tutorial("genius_tutorial", "genius")` in the console.
> 
> @tsoukinator If you're using {spotifyr} already, I'd suggested using the function `get_discography()` i built it for that purpose :P
> 
> Again I'd recommend giving `mirr.netlfiy.app` a perusal. For example there is a handy function definition in there for getting track level audio features without needing the track ID. (https://mirr.netlify.app/audio-features-spotify.html).

Thanks Josiah,

I do recall stumbling upon **get_discography** sometime earlier last year, but did not use it for a variety of reasons:
- Artists with album names which differ between spotify/genius would not have their albums collected (this occurs in the case of Bush - you'll notice "Sixteen Stone" in spotify is named as "Sixteen Stone (Remastered)". I ended up producing my own manual workarounds to this, so dropped get_discography altogether.

As for **get_artist_audio_features** as mentioned in your site - I vaguely recall it not being 100% accurate in the case of finding lesser known artists, who have a name similar to more popular artists. I've had to built my own internal checks for this occurring, and when it does, I resolve via running **get_track_audio_features**. I don't have an example off the top of my head I can re-create on this one. Although if I find it again, I can bring it to your attention. I'm not entirely sure if there's an easy fix for it, and it isn't something I recall happening often.

These were both items I stumbled upon before I became accustomed to posting about them on github. My solution instead of raising them, were to use more granular functions available in your packages.
